### ISSUE TITLE
Browser fails when using UTF-8 code page under windows with outside Ascii-range chars 

### ISSUE THREAD

**Issue by BodoWitcha** [SiCKRAGETV/sickrage-issues#2504]
I have reported a similar issue as #2485 when using code page 850 
This error is caused by the use of characters in the filename outside the normal Ascii range.
In my case `Person.of.Interest.S02E11.2πR.mkv` or
`The.Simpsons.S12E09.HOMЯ.ts`
The characters __π__ and __Я__ are causing the issue.
__miigoto__ suggested to enter the chcp 65001 command, which effectively is telling to use utf-8 (https://msdn.microsoft.com/en-us/library/windows/desktop/dd317756(v=vs.85).aspx)

Changing to code page 65001 (UTF-8) results in a new error:
![image](https://cloud.githubusercontent.com/assets/13832556/9330219/216d5b40-45b8-11e5-8062-a4a93d4d43ee.png)

Because issue #2485 was closed, this new issue created.

**Comment by h3llrais3r** [comment]
As mentioned in issue #2485, I have just tested it from scratch.
If you delete your entire sickbeard.db and start from scratch and you add the shows again, it works. 
To be precise: the gui works but the console outputs a lot of errors:
```
20:01:39 ERROR::Thread-23 :: Unable to get size for file C:\Temp\Hannibal\Season 02\Hannibal.s02e11.Ko.No.Mono.720p.WEB-DL-thumb.jpg Error: error 2 : Het systeem kan het opgegeven bestand niet vinden
Traceback (most recent call last):
  File "C:\Tools\GitHub\SickRage\sickbeard\helpers.py", line 1464, in get_size
    total_size += ek.ek(os.path.getsize, fp)
  File "C:\Tools\GitHub\SickRage\sickbeard\encodingKludge.py", line 68, in ek
    result = func(*args, **kwargs)
  File "C:\Tools\Python-2.7.9\lib\genericpath.py", line 49, in getsize
    return os.stat(filename).st_size
WindowsError: [Error 2] Het systeem kan het opgegeven bestand niet vinden: u'C:\\Temp\\Hannibal\\Season 02\\Hannibal.s02e11.Ko.No.Mono.720p.WEB-DL-thumb.jpg'
20:01:39 ERROR::Thread-23 :: Unable to get size for file C:\Temp\Hannibal\Season 02\Hannibal.s02e11.Ko.No.Mono.720p.WEB-DL.mkv Error: error 2 : Het systeem kan het opgegeven bestand
 niet vinden
Traceback (most recent call last):
  File "C:\Tools\GitHub\SickRage\sickbeard\helpers.py", line 1464, in get_size
    total_size += ek.ek(os.path.getsize, fp)
  File "C:\Tools\GitHub\SickRage\sickbeard\encodingKludge.py", line 68, in ek
    result = func(*args, **kwargs)
  File "C:\Tools\Python-2.7.9\lib\genericpath.py", line 49, in getsize
    return os.stat(filename).st_size
WindowsError: [Error 2] Het systeem kan het opgegeven bestand niet vinden: u'C:\\Temp\\Hannibal\\Season 02\\Hannibal.s02e11.Ko.No.Mono.720p.WEB-DL.mkv'
20:01:39 ERROR::Thread-23 :: Unable to get size for file C:\Temp\Hannibal\Season 02\Hannibal.s02e11.Ko.No.Mono.720p.WEB-DL.nfo Error: error 2 : Het systeem kan het opgegeven bestand niet vinden
Traceback (most recent call last):
  File "C:\Tools\GitHub\SickRage\sickbeard\helpers.py", line 1464, in get_size
    total_size += ek.ek(os.path.getsize, fp)
  File "C:\Tools\GitHub\SickRage\sickbeard\encodingKludge.py", line 68, in ek
    result = func(*args, **kwargs)
  File "C:\Tools\Python-2.7.9\lib\genericpath.py", line 49, in getsize
    return os.stat(filename).st_size
WindowsError: [Error 2] Het systeem kan het opgegeven bestand niet vinden: u'C:\\Temp\\Hannibal\\Season 02\\Hannibal.s02e11.Ko.No.Mono.720p.WEB-DL.nfo'
```

The problem is related to the *Kō* in the filename. The console outputs/searches it as *Ko* so it can't be found...

How can we solve this?

**Comment by BodoWitcha** [comment]
As described in issue #2485 the removal and readding the next 2 shows
* Person of Interest (episode Person.of.Interest.S02E11.2πR.mkv has the non-ASCII character) 
* The.Simpsons (episode The.Simpsons.S12E09.HOMЯ.ts has the non-ASCII character)

circumvented the browser failure. 

The following shows resulted also in the reported issue.
* Æon Flux (here the show name contains the non-ASCII character)
* Without a Trace 
* Tripping the Rift
* The X-Files

For the last 3 shows I could not identify the source of the issue.

The suggested mass update as the next step of the circumvention does __NOT__ remove the issue.

However the removal and re-adding circumvents the (GUI failure) issue, however with error log messages.

I do not know the rules for closing an issue. I suggest as long as an issue is not resolved  it should not be closed. When a circumvention is made available, the priority of finding a solution can be lowered.

A side effect in the manner I have implemented the circumvention is that when a source update is performed the chcp 65001 commend setting is lost. I like to learn how to write a windows command/script which set the UCS-8 code page before starting and restarting SickRage.

**Comment by miigotu** [comment]
Both of the issues are a duplicate of EXISTING issues. These are not new issues.

**Comment by BodoWitcha** [comment]
@miigotu 
Thanks for the explanation, The open non-duplicate one remains open.

**Comment by h3llrais3r** [comment]
And which one is that? Sorry but I can't seem to find it :smile:

**Comment by BodoWitcha** [comment]
Sorry that I have to continue this closed issue, for which I also cannot find an EXISTING other one.

After shutting down the circumvention of the removal and re-adding of the shows is lost.

The command entered and message are shown hereafter.

Then click on the row with the show __Æon Flux__.
The browser shows the following error:
`Error !`

`UnicodeDecodeError: 'utf8' codec can't decode bytes in position 30-31: invalid continuation byte`

`<td align="center"><img src="${sbRoot}/images/${("nfo-no.gif", "nfo.gif")[epResult["hasnfo"]]}" alt="${("N", "Y")[epResult["hasnfo"]]}" width="23" height="11" /></td> <td align="center"><img src="${sbRoot}/images/${("tbn-no.gif", "tbn.gif")[epResult["hastbn"]]}" alt="${("N", "Y")[epResult["hastbn"]]}" width="23" height="11" /></td> <td align="center"> <% if epLoc and show._location and epLoc.lower().startswith(show._location.lower()): epLoc = epLoc[len(show._location)+1:] if epLoc != '' and epLoc != None: text = '<span title="' + epLoc + '" class="addQTip">' + str(epResult['episode']) + "</span>"`
`C:\Python27\lib\encodings\utf_8.py, line 16:
return codecs.utf_8_decode(input, errors, True)`
`c:\SickRage\gui/slick/interfaces/default/displayShow.mako, line 476:`
`<%`
`c:\SickRage\lib\mako\runtime.py, line 853:`
`callable_(context, *args, **kwargs)`

Hereafter back in the browser to the GUI.
Using Manage->Mass update to remove the show __Æon Flux__ and in the Shows overview re-add the __Æon Flux__ show.

Wait until the adding is complete...`(loading)	Loading... (V:\Series\Aeon Flux)`
Meanwhile the command prompt shows the following messages:
`c:\SickRage\lib\imdb\parser\http\utils.py:472: UserWarning: unable to use "lxml": No module named lxml.html`
`  warnings.warn('unable to use "%s": %s' % (mod, str(e)))`
`c:\SickRage\lib\imdb\parser\http\utils.py:463: UserWarning: falling back to "beautifulsoup"
  warnings.warn('falling back to "%s"' % mod)`
`[warn] [/exif/content] [Autofix] Fix parser error: stop parser, found unparsed segment: start 1472, length 16, found unparsed segment: start 1824, length 16`
`[warn] [/exif/content] [Autofix] Fix parser error: stop parser, found unparsed segment: start 1472, length 16, found unparsed segment: start 1824, length 16`
`[warn] [/exif/content] [Autofix] Fix parser error: stop parser, found unparsed segment: start 1472, length 16, found unparsed segment: start 1824, length 16`
`[warn] [/exif/content] [Autofix] Fix parser error: stop parser, found unparsed segment: start 1472, length 16, found unparsed segment: start 1824, length 16`

Then I refreshed the browser.(PF5)
The  show can now be accessed/viewed.
At the end of the browser session a Shutdown is requested.
The command is completed.

When the following commands are re-entered:

`C:\Users\Bob>chcp`
`Active code page: 65001`

`C:\Users\Bob>python c:\SickRage\SickBeard.py -d`

The browser restarts, but opening the show __Aeon Flux__ again result in the same reported issue.

So removing all problem shows and re-adding them when needed is in practice not workable.

I hope that the priority for solving the existing issue is not too low.

**Comment by miigotu** [comment]
-d loses the chcp.

**Comment by BodoWitcha** [comment]
without the -d it appears also looses the chcp

`C:\Users\Bob>chcp 65001`
`Active code page: 65001`

`C:\Users\Bob>python c:\SickRage\SickBeard.py`

After shutdown and restart with 

`C:\Users\Bob>chcp 65001`
`Active code page: 65001`

`C:\Users\Bob>python c:\SickRage\SickBeard.py`

the description is identical, only a lot of messages are shown because it is not --quiet anymore

**Comment by BodoWitcha** [comment]
You are right

**Comment by miigotu** [comment]
If you just use one of the sickrage start scripts that are up to date they should be doing the chcp for you.

**Comment by h3llrais3r** [comment]
I've just tested what @BodoWitcha did and even without the -d, I cannot longer view the show after a shutdown. Only solution is always remove and re-add the show to get it to work... But this is not acceptable.

Next to this, we are still getting the errors in the console. This results in an incorrect file size of your show folder.
```
17:19:33 ERROR::Thread-27 :: Unable to get size for file C:\Temp\Hannibal\Season 02\Hannibal.s02e11.Ko.No.Mono.720p.WEB-DL.mkv Error: error 2 : Het systeem kan het opgegeven bestand niet vinden
Traceback (most recent call last):
  File "C:\Tools\GitHub\SickRage\sickbeard\helpers.py", line 1464, in get_size
    total_size += ek.ek(os.path.getsize, fp)
  File "C:\Tools\GitHub\SickRage\sickbeard\encodingKludge.py", line 68, in ek
    result = func(*args, **kwargs)
  File "C:\Tools\Python-2.7.9\lib\genericpath.py", line 49, in getsize
    return os.stat(filename).st_size
```

**Comment by h3llrais3r** [comment]
@miigotu Could you re-open this issue, because there isn't any duplicate open anymore?

**Comment by BodoWitcha** [comment]
With The Simpsons S12E09 also Unicode encoding problems after every shutdown and start with chcp 65001, this shows breaks with an error.

The update of database SickRage,db looks good:
![image](https://cloud.githubusercontent.com/assets/13832556/9394285/e7aba624-4787-11e5-806a-42486bd551ae.png)

But the browser shows:

Error !
--
`UnicodeEncodeError: 'charmap' codec can't encode character u'\u042f' in position 56: character maps to <undefined>`

`% endif <% curSeason = int(epResult["season"]) %> % endif <% epLoc = str(epResult["location"]) %> <tr class="${Overview.overviewStrings[epCats[epStr]]} season-${curSeason} seasonstyle"> <td class="col-checkbox"> % if int(epResult["status"]) != UNAIRED: <input type="checkbox" class="epCheck" id="${str(epResult["season"])+'x'+str(epResult["episode"])}" name="${str(epResult["season"]) +"x"+str(epResult["episode"])}" /> % endif`

`C:\Python27\lib\encodings\cp1252.py, line 12:`
`return codecs.charmap_encode(input,errors,encoding_table)`

`C:\SickRage\gui/slick/interfaces/default/displayShow.mako, line 466:`
`<% epLoc = str(epResult["location"]) %>`

`C:\SickRage\lib\mako\runtime.py, line 853:`
`callable_(context, *args, **kwargs)`

The  `u'\u042f'` is the __Я__ character.

The message `UnicodeEncodeError: 'charmap' codec can't encode` is odd because [u'\u042f'](http://www.fileformat.info/info/unicode/char/042f/index.htm) is valid unicode character.

**Comment by miigotu** [comment]
I can easily make mako not throw these errors, but the underlying problem will just be masked and you will have a hell of a time trying to figure out "wtf isnt this working?"

This is the start of an ongoing process to move towards completely unicode within SR, so that changing code page is not necessary, and that hidden/undebuggable problems can be fixed.

**Comment by BodoWitcha** [comment]
The unicode cannot encode error was caused by redirecting the output to a file.
Beause the -d option was not allowed as circumvention I have tried the test script:
`set sickragedir="C:\SickRage"`
`chcp 65001`
`python %sickragedir%\SickBeard.py 1>SichRageOutput.log 2>SickRageDebug.log`

this to see the debug output.
But I read on [PyMOTW under the heading Standard Input and Output Streams](http://pymotw.com/2/codecs/) that the mentioned error can error.

So I repeated the test with the altered script:
`C:\SickRage>chcp 65001`
`Active code page: 65001`

`C:\SickRage>python "C:\SickRage"\SickBeard.py  1>nul 2>nul`

Now we get the same error as reported earler by me and @h3llrais3r .

@miigotu - I agree that the target is to cleanup the encoding mess, but because we downloading the master to keep updated we also have to live with an app which is not completely overhauled.

Therefore I can live, and probably many others, with a working circumvention.
So far I was not able, based on the several suggestions, to make a working script as circumvention..

Therefore I can agree with your suggestion to "not throw ... errors".
I am learning as a side effect a little bit about Python and I understood from the explanation on [PyMOTW/codecs](http://pymotw.com/2/codecs/) under Error Handling that instead of  'strict' you can use 'replace' e.g. with █. The result maybe not nice on the screen, but the Gui does not break.

**Comment by miigotu** [comment]
@BodoWitcha maybe you can try something. You will still need the chcp 65001, but stop SR, open sickbeard/webserve.py, and around line 156 change:
`return super(PageTemplate, self).render(*args, **kwargs)`
to 
`return super(PageTemplate, self).render_unicode(*args, **kwargs)`

tell me if that helps

**Comment by miigotu** [comment]
@h3llrais3r you also ^^

**Comment by BodoWitcha** [comment]
As requested the following change applied to sickbeard/webserve.py
![image](https://cloud.githubusercontent.com/assets/13832556/9398388/0b52b832-47a7-11e5-9602-cfd6f3204cb4.png)

After stopping Sr and restarting with the following commands
`C:\SickRage>chcp`
`Active code page: 65001`

`C:\SickRage>python C:\SickRage\SickBeard.py >1nul 2>nul`

it resulted in web error as follows:
![image](https://cloud.githubusercontent.com/assets/13832556/9398415/47bc17e6-47a7-11e5-99a5-fc5af9470ed9.png)

The result is a solution we are not waiting for. Changes removed.

**Comment by BodoWitcha** [comment]
It was late last night when I performed the test with the suggested test.
This afternoon I repeated the test, but I cannot reproduce the reported effect __NameError: Undefined__
The tests today showed that the reported issue is still present
__UnicodeDecodeError:__ 'utf8' codec can't decode bytes in position ...
The suggested change has no visible effect.

**Comment by h3llrais3r** [comment]
Just confirming that the suggested change doesn't work like mentioned by @BodoWitcha. :cry:

**Comment by mbrackeva** [comment]
I gave up and installed a Virtualbox. I now have SR running on a virtual Fedora instance. Runs nicely, faster than on the Windows physical host! The character set is UTF-8, so now no issues with characters anymore.
It's not the first time there are character set issues on Windows after a SR update. Apparently that part isn't properly tested. So if you can, better move away to a supported Linux flavour.

BTW, this issue actually breaks SR! It is more than a nuisance. Are we sure this is treated with the right priority?

**Comment by BodoWitcha** [comment]
With some further analysis I found the following:
Using the debugger in  PyCharm I found that exception was caused by instructions in the file `C:\SickRage\gui\slick\interfaces\default\displayShow.mako`.
First I thought that error was caused by the `str(epResult["location"])` , but it appeared that it was the use of `epLoc` from the assignment `epLoc = str(epResult["location"] `

I have commented out that part in the mako template as shown hereafter:
![image](https://cloud.githubusercontent.com/assets/13832556/9442329/ab21a90a-4a7b-11e5-8e76-047d6ab6bad9.png)

I have no knowledge (yet) what those lines perform, but the reported issue was resolved.

I hope that this information will help in resolving the issue.

**Comment by BodoWitcha** [comment]
In the part, which I have commented out `show._location` contains the path of the show.

                `if epLoc and show._location and epLoc.lower().startswith(show._location.lower()):
                    epLoc = epLoc[len(show._location)+1:]`

which has the following function I think:


IF `eploc` is not '' AND `show._location` is not '' AND the firstpart of `epLoc` is equal `to show._location` THEN
      truncate the firstpart of `epLoc` with the length of `show._location`.
However under Windows, even when chcp 65001 is active, show._location is not a unicode string.
Calculation of len is wrong and the startswith condition will probably not match and as it appears other decoding issues from certain positions.

With my commented out statements  in the template the result of the relevant part in the browser  looks like:
![image](https://cloud.githubusercontent.com/assets/13832556/9450566/ae87abac-4aa9-11e5-930b-92e724ac2607.png)

The same show on my QNAP NAS looks like:
![image](https://cloud.githubusercontent.com/assets/13832556/9450601/e78d5f1e-4aa9-11e5-97ef-6abf76ff5207.png)

I did not yet have found where the `show` variable is set.

I hope that the issue can be resolved soon.

**Comment by BodoWitcha** [comment]
As an extra test I renamed the folder `Aeon Flux` to `Æon Flux`

Under Windows the new location could NOT be found with Add Existing Show.
Changing the missing path using the Edit function alspo failed to accept the selected folder path with the browse option.

Under my QNAP NAS everything worked as expected. This is of course a new issue, but related to this issue.

![image](https://cloud.githubusercontent.com/assets/13832556/9453035/f5d11616-4ab7-11e5-99bd-820c28262a0b.png)

![image](https://cloud.githubusercontent.com/assets/13832556/9453047/085711dc-4ab8-11e5-8e8b-a19c77b5ea03.png)

**Comment by BodoWitcha** [comment]
I found the variable 'show', which I learned is part of the `context` pointing by the `kwargs` dictionary.
Here I found that `show._location` contained the Unicode string `u'W:\\Series\\Lost Girl'`
So my earlier observation that it was not an Unicode string is incorrect, but the issue is caused by the commented out 2 lines somehow.

**Comment by BodoWitcha** [comment]
With the latest master updated of 2015-08-25 (3fb81bd65d7710e59d6c89945a3be39d2dbfc386) (tag v4.0.54) the reported issue for the shows:
* Person of Interest (episode Person.of.Interest.S02E11.2πR.mkv has the non-ASCII character)
The.Simpsons (episode The.Simpsons.S12E09.HOMЯ.mkv has the non-ASCII character)
* Æon Flux (here the show name and file names contains the non-ASCII character)
* Without a Trace
* Tripping the Rift
* The X-Files
* Lost Girl

does not show the reported behavior anymore.

The mentioned behavior on the folder name Æon Flux will be further investigated and reported as an issue when relevant.

**Comment by h3llrais3r** [comment]
Confirming that with the latest version of master, the problem does not occur anymore.
:+1:
